Calculate interpolation weight matrix

bendudson opened the pull request:
Define a matrix operation to calculate the forward and backward interpolation.
This enables matrices to be constructed that represent operators such as `Grad_par` and `Div_par`.

bendudson committed:
weights: Calculate interpolation weights for forward/backward

Create matrix representation of interpolation, saved in CSR format
in the BOUT++ grid file.

bendudson committed:
Apply black/isort changes

dschwoerer commented on zoidberg/weights.py:
```diff
@@ -0,0 +1,221 @@
+                    # Boundary
+                    rows.append(-1)
+                else:
+                    # Not a boundary point => Interpolating
```
```suggestion
                    rows.append(-1)
                    continue
                # Not a boundary point => Interpolating
```
you can avoid some indention with `continue`

dschwoerer reviewed (approved):
This looks good, however that would require quite significant changes on the BOUT++ side as well.

bendudson committed:
weights: Improving forward/backward weights

bendudson committed:
Weights: Add rows in Y boundaries

Needed to enable correct calculation of divergence operator.

bendudson committed:
Weights: LLM makeover

Renaming, some refactoring, more documentation, standard CSR output.

bendudson commented:
Thanks @dschwoerer . BOUT++ changes in this PR: https://github.com/boutproject/BOUT-dev/pull/3330 .
Still a work in progress.

bendudson committed:
make_maps: Add cell sub-sampling

In addition to following a field-line from cell center,
follow a set of points distributed over each cell.

Using Gauss-Legendre points to perform averaging over each cell.

bendudson committed:
Sub-cell field line tracing and mapping

Adds an option to trace multiple field lines per cell (default is 1).
Uses Gauss-Legendre points. The resulting maps are used to calculate
forward & backward matrices.

bendudson committed:
Weights: Major rewrite, calculating C, L+ and L- numberings

Cell center values, forward legs and backward legs are now
three separate 'spaces' with different numberings.
The forward and backward operators are now non-square matrices
that map between spaces.

This complexity is required to handle cells that partly intersect
boundaries: Fluxes are split between the interior and boundary part.
Such cells therefore have two forward (or backward) 'legs' on which
fluxes are calculated.

dschwoerer commented:
I think it might make sense to allow for multiple target BCs for one cell. That might also help with numerical stability, as you do not get such sharp transitions from "sees a target" to "does not see a target". https://github.com/boutproject/BOUT-dev/pull/3308 would allow to implement this in BOUT++, without having to touch the physics code, if they use the new interface.

I am also not sure that https://github.com/boutproject/BOUT-dev/pull/3320 makes sense, if we start computing parallel slices on demand.

bendudson committed:
Weights: Ensure consistency with boundaries

When cells partly intersect boundaries, parallel boundary conditions
need to be applied to the partial boundaries.

bendudson committed:
Apply black/isort changes

bendudson committed:
Merge branch 'master' into calculate-weights

bendudson committed:
Ruff formatting
